python_modules/dagster/dagster/_core/external_execution/task.py

```diff
@@ -57,228 +56,224 @@ def __init__(
         self._tempdir = None
         self._input_path = input_path
         self._output_path = output_path
+        self._socket_server_host = socket_server_host
+        self._socket_server_port = socket_server_port
         self.env = copy(env) if env is not None else {}
 
-    def run(self) -> int:
-        write_target, stdin_fd, input_env_vars = self._prepare_input()
-        read_target, stdout_fd, output_env_vars = self._prepare_output()
-
-        if self._input_mode == ExternalExecutionIOMode.stdio:
-            assert isinstance(write_target, int)
-            input_thread = self._default_input_thread(write_target)
-            input_thread.start()
-        elif self._input_mode == ExternalExecutionIOMode.file:
-            assert isinstance(write_target, str)
-            input_thread = None
-            self._write_input(write_target)
-        elif self._input_mode == ExternalExecutionIOMode.fifo:
-            assert isinstance(write_target, str)
-            input_thread = self._default_input_thread(write_target)
-            input_thread.start()
-        elif self._input_mode == ExternalExecutionIOMode.socket:
-            assert isinstance(write_target, SocketAddress)
-            input_thread = self._start_socket_server(write_target)
-        else:
-            check.failed(f"Unsupported input mode: {self._input_mode}")
+    def run(self) -> None:
+        base_env = {
+            **os.environ,
+            **self.env,
+            DAGSTER_EXTERNAL_ENV_KEYS["input_mode"]: self._input_mode.value,
+            DAGSTER_EXTERNAL_ENV_KEYS["output_mode"]: self._output_mode.value,
+        }
+
+        with ExitStack() as stack:
```

Comment:
Generally find these shenanigans involving an exist stack to be too complicated. can it be refactors to be with blocks?

Reply:
can you elaborate on your concerns with `ExitStack`? 

I think conditionally entering contexts like below for socket mode is annoying / leads to extra indentation without it 

I like `ExitStack` here

Reply:
ya i think there is a difference between `ExitStack` used via `with` which is clean and a manually book kept one which i agree is gnarlier.

Reply:
I think `ExitStack` is cleanest here because:

- we are conditionally opening at least one context manager, which is painful using `with ...`
- there are four context managers open so blocks would be deeply nested

Reply:
Basically I think it should be the "tool of last resort" because I think code paths with `ExitStack` are more likely to aggressive atrophy as other people wonder in and out of this section of the codebase. So my general is show my the alternative and if it is truly horrible then consider `ExitStack`

Reply:
ok I'll give it a shot